### PR TITLE
Is765/product group and vendor (🗃️ DB change) 

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

- ✨ New optional ``products.vendor.invitation_url``
- ✨ ``services/webserver`` API  ``0.13.1 → 0.14.0``
   - ``GET /groups`` returns an item with ``product`` group 
   - ``product.group_id`` returned as well via ``/statics``
 - ✨ event to ensure all products have an associated group-id

### 🗃️ DB change
 - MINOR: adds an optional ``products.vendor.invitation_url`` ask @odeimaiz for config upon deploy

## Related issue/s

- Implements parts of ITISFoundation/osparc-issues#765
  - ``Include s4l-lite registered users in product-specific everyone organization`` 
     -  ``API: return ``product`` group``
  - ``Invitations``
     -  ``Add how to get an invitation link to product vendor info in DB. S4L-Lite should point to the zmt website form``

## How to test
- ``product.group_id=12``:
image
- This is with ``user_id=1``
image
- This is with ``user_id=9``
```json
{
  "data": {
    "me": {
      "gid": 11,
      "label": "crespo",
      "description": "primary group",
      "thumbnail": null,
      "accessRights": {
        "read": true,
        "write": false,
        "delete": false
      },
      "inclusionRules": {}
    },
    "organizations": [
      {
        "gid": 12,
        "label": "s4l",
        "description": "s4l product group",
        "thumbnail": null,
        "accessRights": {
          "read": true,
          "write": false,
          "delete": false
        },
        "inclusionRules": {}
      }
    ],
    "all": {
      "gid": 1,
      "label": "Everyone",
      "description": "all users",
      "thumbnail": null,
      "accessRights": {
        "read": true,
        "write": true,
        "delete": false
      },
      "inclusionRules": {}
    },
    "product": {
      "gid": 12,
      "label": "s4l",
      "description": "s4l product group",
      "thumbnail": null,
      "accessRights": {
        "read": true,
        "write": false,
        "delete": false
      },
      "inclusionRules": {}
    }
  },
  "error": null
}
```

## Checklist

- [x] ``make version-*``
- [x] ``make openapi.json``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/groups_handlers.py [hunk 1]
```diff
@@ -59,16 +59,39 @@ async def list_groups(request: web.Request):
...
+            )
+        except GroupNotFoundError as err:
+            logger.debug(
+                "This user does not belong to any product's group: %s."
+                "This is typically assigned during registration but this user "
+                "might have been created prior to this new feature."
+                "TIP: assign it manually to the default product group_id and "
+                "restart service to refresh product info",
+                err,
+            )
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/groups_handlers.py:91]: I think this message will never be seen in production environments nor anybody will act upon it.

File: services/web/server/src/simcore_service_webserver/products_events.py [hunk 2]
```diff
@@ -33,6 +36,27 @@ async def setup_product_templates(app: web.Application):
...
+    async with engine.acquire() as conn:
+        async for row in iter_products(conn):
+            product_name = row.name
+            product_group_id = await get_or_create_product_group(conn, product_name)
+            log.debug(
+                "Missing group for %s, created %s",
+                f"{product_name=}",
+                f"{product_group_id=}",
+            )
+
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/products_events.py:58]: Please consider this might run in parallel, since we have multiple instances of the webserver. Can this cause issues?

**Review by GitHK — COMMENTED** [review]
I have some questions, please see them below.

File: services/web/server/src/simcore_service_webserver/groups_handlers.py (hunk 1 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/groups_handlers.py:91]: True. The idea is that if user does not belong to any product group, then None is returned

**Review by odeimaiz — APPROVED** [review]
👍

File: services/web/server/src/simcore_service_webserver/products_events.py (hunk 2 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/products_events.py]: Good point! I made some changes in ``get_or_create_product_group `` to achieve this and added a test

File: packages/postgres-database/src/simcore_postgres_database/utils_products.py [hunk 3]
```diff
@@ -63,25 +63,30 @@ async def get_or_create_product_group(
     """
     Returns group_id of a product. Creates it if undefined
     """
-    group_id = await get_product_group_id(connection, product_name=product_name)
-    if group_id is not None:
-        return group_id
-
     async with connection.begin():
```
Comment (GitHK) [inline packages/postgres-database/src/simcore_postgres_database/utils_products.py:66]: This looks like a transaction, in that case all the rest looks good to me.

File: packages/postgres-database/tests/products/test_utils_products.py [hunk 4]
```diff
@@ -97,3 +98,30 @@ async def test_get_or_create_group_product(
...
+                    products.c.priority
+                )
+            ):
+                # get or create
+                product_group_id = await get_or_create_product_group(
+                    conn, product_name=product_row.name
+                )
+                return product_group_id
+
+    tasks = [asyncio.create_task(_auto_create_products_groups()) for _ in range(5)]
```
Comment (GitHK) [inline packages/postgres-database/tests/products/test_utils_products.py:123]: Maybe having this very high, will point out possible issues, and if it works like that it should always be fine.
```suggestion
    tasks = [asyncio.create_task(_auto_create_products_groups()) for _ in range(100)]
```

**Review by GitHK — APPROVED** [review]
👍 Looks good to me now!

**Review by odeimaiz — CHANGES_REQUESTED** [review]
Please, provide also product-specific everyone organization in the /me call

**Review by odeimaiz — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/groups_api.py [hunk 5]
```diff
@@ -104,13 +111,32 @@ async def _get_user_from_email(app: web.Application, email: str) -> RowProxy:
...
+async def get_product_group_for_user(
+    app: web.Application, user_id: int, product_gid: int
+) -> dict[str, str]:
+    """
+    Returns product's group if user belongs to it, otherwise it
+    raises GroupNotFoundError
+    """
+    engine = app[APP_DB_ENGINE_KEY]
+    async with engine.acquire() as conn:
+        group: RowProxy = await _get_user_group(conn, user_id, product_gid)
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/groups_api.py:136]: Q: Can a user be part of several product groups?

File: services/web/server/src/simcore_service_webserver/groups_handlers.py [hunk 6]
```diff
@@ -59,16 +60,31 @@ async def list_groups(request: web.Request):
 
     List of the groups I belonged to
     """
+    from .products import Product, get_current_product
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/groups_handlers.py:63]: why is this import here?

**Review by sanderegg — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/groups_api.py (hunk 5 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/groups_api.py:136]: yes

File: services/web/server/src/simcore_service_webserver/groups_handlers.py (hunk 6 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/groups_handlers.py:63]: this is because the routes are collected using an old map function ``get_handlers_from_namespace(groups_handlers)`` that "thinks" that ``get_current_product`` is a handler. All theis function is deprecated and will go away when we have removed opeanpi-core
